lib/src/Base/Stat/HMatrixImplementation.cxx

```diff
@@ -320,19 +348,40 @@ void HMatrixImplementation::factorize(const String& method)
   else if (method != "LU")
     LOGWARN( OSS() << "Unknown factorization method: " << method << ". Valid values are: LU, LDLt, or LLt.");
 
-  try
-  {
-    hmat_factorization_context_t context;
-    hmat_factorization_context_init(&context);
-    context.factorization = fact_method;
-    context.progress = NULL;
-    static_cast<hmat_interface_t*>(hmatInterface_)->factorize_generic(static_cast<hmat_matrix_t*>(hmat_), &context);
-  }
-  catch (std::exception& ex)
-  {
-    // hmat::LapackException is not yet exported
-    throw InternalException(HERE) << ex.what();
-  }
+  // Compute a reasonable regularization factor
+  Scalar lambda = 2.0 * computeApproximateLargestEigenValue() * ResourceMap::GetAsScalar("HMatrix-AssemblyEpsilon");
+  // Do regularization
+  addIdentity(lambda);
+  Bool done = false;
+  String msg;
+  const UnsignedInteger maximumIteration = ResourceMap::GetAsUnsignedInteger("HMatrix-FactorizationIterations");
+  for (UnsignedInteger iteration = 0; iteration < maximumIteration; ++ iteration)
+    {
+      LOGDEBUG(OSS() << "Factorization, regularization loop " << iteration << ", regularization factor=" << lambda);
+      try
+	{
+	  hmat_factorization_context_t context;
+	  hmat_factorization_context_init(&context);
+	  context.factorization = fact_method;
+	  context.progress = NULL;
+	  static_cast<hmat_interface_t*>(hmatInterface_)->factorize_generic(static_cast<hmat_matrix_t*>(hmat_), &context);
+	  done = true;
+	  LOGDEBUG("Factorization ok");
+	}
+      catch (std::exception& ex)
+	{
+	  // hmat::LapackException is not yet exported
+	  msg = ex.what();
+	  // Double the current regularization factor by adding it another time
+	  addIdentity(lambda);
+	  // And double its value for next loop
+	  lambda += lambda;
```

Comment:
ok for 2 but I think this value should be in a way compared to CovModel(0) ?
@poncetr ?

Reply:
I don't think so as the initial value for lambda is the eigenvalue of the covariance matrix, ie it is already in the correct scale.

Reply:
If we consider a `CorrelationModel` instead of `CovarianceModel` (`amplitude` set to 1),
we might easily have maximal eigen value around 1000 if we consider medium/large sized matrices.
Thus adding `1` to the diagonal for regularising the matrix is statistically "incorrect" as we are changing the covariance structure.

Reply:
I still don't understand where you see a 1 being added to the diagonal. Tell me where I am wrong:
  + I want to factorize a matrix obtained through the discretization of a given covariance model, regardless of its scaling (ie covariance or correlation)
  + To this end I first compress it using an h-matrix representation. It introduces a numerical noise which can break the definite positiveness of the matrix
  + To fix this problem I look for the optimal scaling of the regularization: something which allows me to shift all the eigenvalues in the positive side. It is a proportion of the largest eigenvalue, modulated by the amount of numerical noise generated by the compression

There is no scaling done between the assembly and the factorization of the matrix. Even if you scaled it before or if you scale it after the factorization, if the factorization succeeded it is ok for the other computations.

Reply:
Agreed. IMO you should also control the regularization : `2 * eps * lamba_max` is ok but with the loop we should assert that final `translating` parameter is less or equal than `0.5 * amplitude`

We should control positiveness and "behavioural" aspects

Reply:
There is definitely additional work to do to ensure that the global problem is well-posed, but IMO:
  + it is well beyond the scope of this PR (eg: is the mesh adapted to the sampling of the covariance model? In particular, do we have clusters of vertices much denser than the scale of the model?)
  + this PR already improves a lot wrt the current implementation, which is essentially useless for the user -> this point alone should be enough to adopt this PR!

BTW, a control based on the eigenvalue takes into account both the amplitude of the model (linked to the output) and the geometry of the sampling points (linked to the scale, ie the input), so the final check should incorporate both aspects. I will add something like this. At the end of the regularization loop, which should end after1 or 2 iterations (not taking into account round-off errors), either the factorization is done, which is a certificate for positiveness, or an exception is thrown, allowing the user to take an action. Could you elaborate a little bit more on the behavioural aspects? I don't see the mathematics beyond this sentence ;-)!

Reply:
I confirm the sclaing of the regularization, as explained by Regis.

From our tests we have observed a linear relationship between the prescribed assembly error of the HMatrix (set by the key ```HMatrix-AssemblyEpsilon``` of the ```ResourceMap```). The regression coefficient is in practice close to one.

If a much larger coefficient is required to ensure the positivity of the approximation of the covariance matrix, then it is likely that the HMatrix library
failed to satisfy the prescribed accuracy, which may especially happen with the "ACA partial" compression method (set by the key ```HMatrix-CompressionMethod``` of the '''ResourceMap```). In this case, it is indeed necessary to abort the computation and suggest to the user to use a more reliable compression algorithm.

FYI, a randomized ACA algorithm is being tested to prevent these possible problems, with the same complexity as the "ACA partial" algorithm.